doc/devguide/extending/detect/app-layer-frames.rst

```diff
@@ -0,0 +1,173 @@
+
+    alert tcp any any -> any any (flow:to_server; content:"|ff|SMB"; offset:4; depth:4;
+    \content:"vigilant meerkat"; sid:1; rev:1;)
+
+For a rule like the above, even though one is interested in *SMB* traffic only, Suricata will inspect **all** *TCP* traffic to the back-end - regardless of application layer protocols.
+
+What the Stream Frames offer is the ability to "point" to a specific portion of stream and identify what type of traffic Suricata is looking at.Then, as the engine reassembles the stream, one can have "read access" to that portion of the stream, aggregating concepts like: what type of application layer protocol that is, and differentiating between `header`, `data` or even protocol versions (`SMB1`, `SMB2`...).
+
+The goal of the stream *Frames* is to expose application layer protocol `PDUs <https://en.wikipedia.org/wiki/Protocol_data_unit>`_ and other such arbitrary elements to the detection engine directly, instead of relying on Transactions. The main purpose is to bring *TCP.data* processing times down by specialising/ filtering down traffic detection.
+
+Benefits for Rule Writers
+=========================
+
+With frame support, the previous example rule becomes something like::
+
+    alert smb any any -> any any (flow:to_server; frame:smb.smb1.data; content:"vigilant meerkat";
+    \sid:1; rev:1;)
+
+As this hypothetical rule shows, this allows for easier to write, more human-readable rules, with the benefit that they are also less costly for the engine to inspect.
+
+Another advantage of this feature is that there is a single rule keyword for application layer frames: ``frame``.
+This keyword takes an argument to specify the per protocol frame type::
+
+    frame:<app proto name>.<specific frame name>
+
+Examples::
+
+    tls.pdu
+    smb.nbss.pdu
+    smb.smb1.pdu
+
+Rules will look something like this::
+
+    alert tls any any -> any any (flow:to_client; frame:tls.pdu; \
+         content:"|17 03 03|"; startswith; sid:1;)
+    alert smb any any -> any any (flow:to_server; frame:smb.nbss.pdu; \
+         content:"|fe|SMB"; offset:4; depth:4; sid:2;)
+    alert smb any any -> any any (flow:to_server; frame:smb.smb1.pdu; content:"|ff|SMB"; sid:3;)
+
+
+Implementation & API Callbacks
+==============================
+
+The application layer parser exposes frames it supports to the detect engine, by tagging them as they're parsed. The rest works automatically.
+
+A code example from the `SMB` decoder - code excerpt from `PR #6713 <https://github.com/OISF/suricata/pull/6713/commits/88c1fea77f55284c74443eaaa230e0c21956433f>`_:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: pub fn parse_tcp_data_ts<'b>
+    :end-at: nbss_hdr.length as i64, SMBFrameType::NBSSData as u8);
+    :lines: 1-3, 56-64
+    :dedent: 4
```

Comment:
Not sure if this is the best option, for this specific code-snippet. If we cut portions of the code, the snippet looks a bit weird. But if we don't, there are just too many lines...